### DISCUSSION TITLE
Getting an output from the integrator

### DISCUSSION THREAD

**Discussion by OliviaBorgstrom** [mitsuba-renderer/mitsuba3#1068] [Q&A]
I'm modifying the ptracer integrator for my own purpose, trying specifically to output positions of hits on a sensor as an array. The biggest difficulty for me here is the outputting part, as I'm finding it quite hard to trace back through the call stack to output something back to python. Is there any examples in the codebase I can look at to help me? 
 Any kind of pointer is useful! Thanks :)

**Comment by njroussel** [comment]
Hi @OliviaBorgstrom 

I don't think we have an example of this.

If you're doing this in a `scalar` variant, it should be fairly straightforward: at every `block->put()` call in the `ptracer`, you can write the position to some `std::vector` with a lock.
With a JIT-enabled variant this is a bit more difficult. You'll need to scatter the position to some pre-allocated buffer. You can either figure out some indexing scheme to not overwrite positions of other threads, or you can use an atomic counter that you re-purpose as an index (see [`dr.scatter_inc`](https://drjit.readthedocs.io/en/latest/reference.html). Finally, you'll need to access that buffer of positions from Python, you can create a new method just for that, and bind it in Python [here](https://github.com/mitsuba-renderer/mitsuba3/blob/master/src/render/python/integrator_v.cpp#L391-L436).
